11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+			m_kernelNormalizedSpectrums[0]->getCreationParameters().image->setObjectDebugName("Kernel red channel spectrum");
+			m_kernelNormalizedSpectrums[1]->setObjectDebugName("Kernel green channel spectrum view");
+			m_kernelNormalizedSpectrums[1]->getCreationParameters().image->setObjectDebugName("Kernel green channel spectrum");
+			m_kernelNormalizedSpectrums[2]->setObjectDebugName("Kernel blue channel spectrum view");
+			m_kernelNormalizedSpectrums[2]->getCreationParameters().image->setObjectDebugName("Kernel blue channel spectrum");
+
+			// Invoke a workgroup per two vertical scanlines. Kernel is square and runs first in the y-direction.
+			// That means we have to create a shader that does an FFT of size `paddedKerDim.height = paddedKerDim.width` (length of each column, already padded to PoT), 
+			// and call `paddedKerDim.width / 2` workgroups to run it. We also have to keep in mind `paddedKerDim.y = WorkgroupSize * ElementsPerInvocation`. 
+			// We prefer to go with 2 elements per invocation and max out WorkgroupSize when possible.
+			// This is because we use PreloadedAccessors which reduce global memory accesses at the cost of decreasing occupancy with increasing ElementsPerInvocation
+
+			// Create descriptor sets
+			const IGPUDescriptorSetLayout* kernelDSLayouts[2] = { kernelFirstAxisFFTPipelineLayout->getDescriptorSetLayout(0), kernelNormalizationPipelineLayout->getDescriptorSetLayout(0) };
+			smart_refctd_ptr<IDescriptorPool> kernelFFTDSPool = m_device->createDescriptorPoolForDSLayouts(IDescriptorPool::ECF_NONE, { kernelDSLayouts, 2 });
+			smart_refctd_ptr<IGPUDescriptorSet> kernelFFTDescriptorSets[2];
+			uint32_t dsCreated = kernelFFTDSPool->createDescriptorSets({kernelDSLayouts, 2}, kernelFFTDescriptorSets);
+
+			// Cba to handle errors it's an example
+			if (dsCreated != 2)
+				return logFail("Failed to create Descriptor Sets!\n");
+
+			// Write descriptor sets
+			updateDescriptorSetFirstAxisFFT(kernelFFTDescriptorSets[0].get(), m_kerImageView);
+			updateDescriptorSetNormalization(kernelFFTDescriptorSets[1].get(), m_kernelNormalizedSpectrums);
+
+			// Compute required WorkgroupSize and ElementsPerThread for FFT
+			// Remember we assume kernel is square!
+			uint32_t maxWorkgroupSize = *m_device->getPhysicalDevice()->getLimits().maxWorkgroupSize;
+			uint32_t elementsPerThread = 1, workgroupSize;
+			do
+			{
+				elementsPerThread <<= 1;
+				workgroupSize = paddedKerDim.width / elementsPerThread;
+			} 
+			while (workgroupSize > maxWorkgroupSize);
+
+			// Create shaders
+			smart_refctd_ptr<IGPUShader> shaders[3];
+			shaders[0] = createShader("app_resources/kernel_fft_first_axis.hlsl", workgroupSize, elementsPerThread, bloomScale);
+			shaders[1] = createShader("app_resources/kernel_fft_second_axis.hlsl", workgroupSize, elementsPerThread, bloomScale);
+			shaders[2] = createShader("app_resources/kernel_spectrum_normalize.hlsl", workgroupSize, elementsPerThread, bloomScale);
```

Comment:
why is the bloomScale being passed into the shader creation!?

Reply:
bloom scale (see `ditt` branch denoiser) is mediated by doing a linear interpolation between a dirac delta function (phase shifted ofc) and the spectrum of the original kernel, WHILE doing the convolution with the main image

Reply:
I'm still a bit confused about this scale thing. I get we can dynamically interpolate between a dirac delta's spectrum and the kernel's spectrum, but is this dynamic sharpening or dynamic scaling? Because the bloomScale being passed here is an argument to determine the size of the kernel before computing its FFT (it's used to create a minified version of the kernel)

Reply:
you can't minify the kernel more than 50% otherwise you'd have to tap the 2nd mipmap (index 1) and we can't compute those right now, you can enlarge as much as you want though

The whole scale thing happening here is just to prevent extreme scaling in the frequency domain requiring an expensive interpolation filter (mitchell or lanczos) to prevent cross-ringing of the bloom star